### PR TITLE
Report verification and recovery state to posthog

### PR DESCRIPTION
**Pull request by BillCarsonFr**
See https://github.com/element-hq/crypto-internal/issues/226

Updating analytics-event to 0.21.0
The reporting to posthog is added in the `DeviceListener` where there is already logic to check state to prompt to verify or setup 4S.

This reports a user property as well as status change event when the verification state or the recovery state changes:
https://github.com/matrix-org/matrix-analytics-events/blob/fd0fe7894a8d47497d87bb90b985f6094447ef77/schemas/CryptoSessionStateChange.json#L3
## Checklist

-   [ ] Tests written for new code (and old code if feasible).
-   [ ] New or updated `public`/`exported` symbols have accurate [TSDoc](https://tsdoc.org/) documentation.
-   [ ] Linter and other CI checks pass.
-   [ ] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md)).

### CODE REVIEW

**Comment by richvdh** [comment]
> See [element-hq/crypto-internal#226](https://github.com/element-hq/crypto-internal/issues/226)

This says:

> The goal is to get an idea of how many existing EX sessions will get asked to re-verify if we force verification.

... so I don't think this change is necessary?

**Comment by BillCarsonFr** [comment]
> > See [element-hq/crypto-internal#226](https://github.com/element-hq/crypto-internal/issues/226)
> 
> This says:
> 
> > The goal is to get an idea of how many existing EX sessions will get asked to re-verify if we force verification.
> 
> ... so I don't think this change is necessary?

The issue specifically says **I'd like to know the numbers for EX and EW.**

**Comment by richvdh** [comment]
> The issue specifically says I'd like to know the numbers for EX and EW.

The issue seems contradictory, but ok.

File: src/DeviceListener.ts [hunk 1]
```diff
@@ -407,6 +415,72 @@ export default class DeviceListener {
...
+                // if they are not, the backup is incomplete
+                const backupPrivateKeyIsInCache = (await crypto.getSessionBackupPrivateKey()) != null;
+                if (secretStorageReady && allCrossSigningSecretsCached && backupPrivateKeyIsInCache) {
+                    recoveryState = "Enabled";
+                } else {
+                    recoveryState = "Incomplete";
+                }
+            } else {
+                // no backup just consider cross-signing secrets
+                if (allCrossSigningSecretsCached && crossSigningStatus.privateKeysInSecretStorage) {
```
Comment (richvdh) [inline src/DeviceListener.ts:458]: could we make this more consistent with the check when there is a backup?

```suggestion
                if (secretStorageReady && allCrossSigningSecretsCached) {
```

(I think the two are equivalent?)

File: src/DeviceListener.ts [hunk 2]
```diff
@@ -407,6 +415,72 @@ export default class DeviceListener {
...
+     * @private
+     */
+    private async reportCryptoSessionStateToAnalytics(
+        crypto: CryptoApi,
+        cli: MatrixClient,
+        secretStorageReady: boolean,
+    ): Promise<void> {
+        //
+        const crossSigningStatus = await crypto.getCrossSigningStatus();
+        // const backupInfo = await crypto.getActiveSessionBackupVersion();
```
Comment (richvdh) [inline src/DeviceListener.ts:433]:
```suggestion
```

File: src/DeviceListener.ts [hunk 3]
```diff
@@ -407,6 +415,72 @@ export default class DeviceListener {
...
+     * @param cli - the matrix client
+     * @param secretStorageReady - if the secret storage is ready
+     * @private
+     */
+    private async reportCryptoSessionStateToAnalytics(
+        crypto: CryptoApi,
+        cli: MatrixClient,
+        secretStorageReady: boolean,
+    ): Promise<void> {
+        //
```
Comment (richvdh) [inline src/DeviceListener.ts:431]:
```suggestion
```

**Review by richvdh — CHANGES_REQUESTED** [review]

File: src/DeviceListener.ts [hunk 4]
```diff
@@ -407,6 +415,72 @@ export default class DeviceListener {
...
+    /**
+     * Reports current recovery state to analytics.
+     * Checks if the session is verified and if the recovery is correctly setup (i.e all secrets known locally and in 4S).
+     * @param crypto - the crypto module
+     * @param cli - the matrix client
+     * @param secretStorageReady - if the secret storage is ready
+     * @private
+     */
+    private async reportCryptoSessionStateToAnalytics(
+        crypto: CryptoApi,
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/DeviceListener.ts]: done

File: src/DeviceListener.ts [hunk 5]
```diff
@@ -407,6 +415,72 @@ export default class DeviceListener {
...
+        cli: MatrixClient,
+        secretStorageReady: boolean,
+    ): Promise<void> {
+        //
+        const crossSigningStatus = await crypto.getCrossSigningStatus();
+        // const backupInfo = await crypto.getActiveSessionBackupVersion();
+        const backupInfo = await this.getKeyBackupInfo();
+        const is4SEnabled = (await cli.secretStorage.getDefaultKeyId()) != null;
+
+        const verificationState = crossSigningStatus.publicKeysOnDevice ? "Verified" : "NotVerified";
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/DeviceListener.ts]: Right, we want to know if the device is actually signed.
So I switch from using `CrossSigningStatus`, and now I use the `DeviceVerificationStatus`. Using `signedByOwner` and `crossSigningVerified`

File: src/DeviceListener.ts (hunk 1 above)
Comment (BillCarsonFr) [reply src/DeviceListener.ts]: Yes there is no backup so `secretStorageReady` checks for cross signing secrets to be in 4S.
Updated and added a new test

File: src/DeviceListener.ts [hunk 6]
```diff
@@ -417,24 +416,22 @@ export default class DeviceListener {
...
+        const secretStorageReady = await crypto.isSecretStorageReady();
         const crossSigningStatus = await crypto.getCrossSigningStatus();
-        // const backupInfo = await crypto.getActiveSessionBackupVersion();
         const backupInfo = await this.getKeyBackupInfo();
         const is4SEnabled = (await cli.secretStorage.getDefaultKeyId()) != null;
+        const deviceVerificationStatus = await crypto.getDeviceVerificationStatus(cli.getUserId()!, cli.getDeviceId()!);
 
-        const verificationState = crossSigningStatus.publicKeysOnDevice ? "Verified" : "NotVerified";
+        const verificationState =
+            !!deviceVerificationStatus?.signedByOwner && !!deviceVerificationStatus?.crossSigningVerified
```
Comment (richvdh) [inline src/DeviceListener.ts:432]: `!!` seems redundant here, given it's in a boolean expression.

**Review by richvdh — APPROVED** [review]
LGTM otherwise

**Review by MidhunSureshR — APPROVED** [review]

**Pull request merged**